src/algorithms/models/vertex.ts

```diff
@@ -40,17 +40,72 @@ class Vertex {
    * Counted as the number of neighbors.
   */
   get outDegree(): number {
-    return this.neighbors.length;
+    return this.neighbors.size;
   }
 
   /**
    * Add a neighbor with edge
    */
   addNeighbor(costs: EdgeCost, vertex: Vertex): void {
-    this.neighbors.push({
+    this.neighbors.add({
       costs,
       vertex
     });
+    vertex.inDegree += 1;
+  }
+
+  /**
+   * Remove a neighbor of this vertex
+   * Outcoming edge of h=this vertex has to be passed as a parameter. The edge that matches the passed one will be
+   * removed from the neighbors array, therefore neighbor is removed.
+   * @param edge Edge to a neighbor that will be removed
+   */
+  removeNeighbor(edge: {vertex: Vertex, costs: EdgeCost}): void {
+    this.neighbors.delete(edge);
+    edge.vertex.inDegree -= 1;
+  }
+
+  /**
+   * Method verifies whether there is an edge back from the 'from' vertex back to this vertex
+   * If there is at least one edge from 'from' vertex' to this vertex true is returned. Otherwise false.
+   * @param from : Vertex Vertex to get back from
+   * @return boolean
+   */
+  canGetBack(from: Vertex): boolean {
+    for (const edge of from.neighbors) {
+      if (edge.vertex === this) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Method verifies whether the inDegree and outDegree of the vertex is the same and whether it is equal to the degree
+   * True is returned when both degrees are equal to the degree parameter. Otherwise false.
+   * @param degree : number outDegree and inDegree that should be equal to
+   * @return boolean
+   */
+  equalDegrees(degree: number): boolean {
+    if (this.inDegree !== this.outDegree) {
+      return false;
+    }
+    return this.outDegree === degree;
+  }
+
+  /**
+   * Get the first edge that leads to other vertex than notThisVertex
+   * This is useful when you are sure that there are only two edges leaving the vertex.
+   * @param notThisVertex : Vertex Edge to this vertex will not be returned
+   * @return object
+   */
+  getEdgeOtherNeighbor(notThisVertex: Vertex): {vertex: Vertex, costs: EdgeCost} {
```

Comment:
I don't like the fact that the return value is some object but sometimes even null is returned. I think we should have a class for an Edge too, with two properties vertex and costs, because we use it in a lot of places so it would be easier to manipulate with them.
Other way could be adding another return type, but I'd prefer changing this function in a way that null doesn't have to be returned at all (as I mentioned in a comment by the _filterVertexNeighbors_ function).
```suggestion
  getEdgeOtherNeighbor(notThisVertex: Vertex): {vertex: Vertex, costs: EdgeCost} | null {
```

Reply:
We are definitely gonna change that as you propose. There is also a need for the list of bypassed nodes. I think we should close all the PRs before making that change.